cp1616/src/cp1616_io_controller.cpp

```diff
@@ -0,0 +1,510 @@
+/*********************************************************************************************//**
+* @file cp1616_io_controller.cpp
+* 
+* cp1616_io_controller class
+* 
+* Copyright {2015} {Frantisek Durovsky}
+* 
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+
+*      http://www.apache.org/licenses/LICENSE-2.0
+
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+   
+* *********************************************************************************************/
+#ifndef CP1616_IO_CONTROLLER_CPP
+#define CP1616_IO_CONTROLLER_CPP
+
+#include <cp1616/cp1616_io_controller.h>
+#include <cp1616/cp1616_io_controller_callbacks.h>
+
+Cp1616IOController *Cp1616IOController::controller_instance_ = 0;
+
+Cp1616IOController* Cp1616IOController::getControllerInstance()
+{
+  if( !controller_instance_ )
+  {
+    controller_instance_ = new Cp1616IOController();
+  }
+  return controller_instance_;
+}
+
+Cp1616IOController::Cp1616IOController() :
+  cp_ready_(0),
+  sem_mod_change_(0),
+  cp_id_(1),
+  cp_handle_(0),
+  cp_current_mode_(PNIO_MODE_OFFLINE),
+  cp_local_state_(PNIO_S_GOOD),
+  device_input_count_(0),
+  device_output_count_(0),
+  total_input_size_(0),
+  total_output_size_(0)
+{
+  //Allocate memory for Input module
+  device_input_length_  = new PNIO_UINT32 [NUM_OF_INPUT_MODULES * sizeof(PNIO_UINT32)];
+  device_input_state_   = new PNIO_IOXS /*volatile*/ [NUM_OF_INPUT_MODULES * sizeof(PNIO_IOXS)];
+  device_input_address_ = new PNIO_ADDR [NUM_OF_INPUT_MODULES * sizeof(PNIO_ADDR)];
+
+  //Allocate memory for Output module
+  device_output_length_  = new PNIO_UINT32 [NUM_OF_OUTPUT_MODULES * sizeof(PNIO_UINT32)];
+  device_output_state_   = new PNIO_IOXS /*volatile*/ [NUM_OF_OUTPUT_MODULES * sizeof(PNIO_IOXS)];
+  device_output_address_ = new PNIO_ADDR [NUM_OF_OUTPUT_MODULES * sizeof(PNIO_ADDR)];
+}
+
+Cp1616IOController::~Cp1616IOController()
+{
+  if(device_input_count_ > 0)
```

Comment:
There is the potential for a memory leak here.  You should delete the pointers no matter what.